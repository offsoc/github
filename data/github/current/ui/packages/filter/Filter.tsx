import {type TestIdProps, testIdProps} from '@github-ui/test-id-props'
import {Box, type SxProp, useResponsiveValue} from '@primer/react'
import type {ResponsiveValue} from '@primer/react/lib-esm/hooks/useResponsiveValue'
import type React from 'react'
import {useCallback, useState} from 'react'

import {AdvancedFilterDialog, type FilterButtonVariant} from './advanced-filter-dialog/AdvancedFilterDialog'
import {ClearButton} from './clear-button/ClearButton'
import {defaultFilterSettings} from './constants/defaults'
import {FilterContextProvider} from './context/RootContext'
import {FilterQuery} from './filter-query'
import {FilterInputIcon} from './FilterInputIcon'
import {FilterInputWrapper} from './FilterInputWrapper'
import {Input} from './input/Input'
import {SubmitButton} from './input/SubmitButton'
import {SuggestionsList} from './suggestions/SuggestionsList'
import type {FilterProvider, FilterSettings, FilterVariant, SubmitEvent} from './types'
import {ValidationMessage} from './ValidationMessage'

// Exports from the root of the package
export * from './constants/defaults'
export * from './filter-query'
export * from './FilterRevert'
export * from './types'
export * from './ValidationMessage'

const filterInputWrapperStyles = ({variant, onSubmit}: {variant: FilterProps['variant']; onSubmit: boolean}) => {
  const leftRounded = variant === 'input'
  const rightRounded = (variant === 'input' && !onSubmit) || (variant !== 'input' && !onSubmit)

  return {
    display: 'grid',
    gridTemplateColumns: onSubmit ? '1fr min-content' : 'min-content 1fr min-content',
    alignItems: 'center',
    pl: onSubmit ? '12px' : 1,
    pr: '1px',
    width: '100%',
    borderStyle: 'solid',
    borderWidth: '1px',
    borderColor: 'border.default',
    marginInlineEnd: -1,
    ...(leftRounded && {
      borderTopLeftRadius: 2,
      borderBottomLeftRadius: 2,
    }),
    ...(rightRounded && {
      borderTopRightRadius: 2,
      borderBottomRightRadius: 2,
    }),
    ':has(input:focus-visible)': {
      zIndex: 1,
      outline: '2px solid var(--borderColor-accent-emphasis, var(--color-accent-emphasis))',
      outlineOffset: '-2px',
    },
  }
}

export type FilterProps = {
  /** ID of the Filter Component */
  id: string
  /** Change event callback, passing the updated filter value */
  onChange?: (value: string) => void
  /** Event callback, triggered by the filter query being parsed, and passing the updated filter query request*/
  onParse?: (request: FilterQuery) => void
  /** Submit event callback, triggered by Enter or pressing Submit, and passing the updated filter query and the
   * originating event type */
  onSubmit?: (request: FilterQuery, eventType: SubmitEvent) => void
  /** Array of FilterProviders that are being used/supported */
  providers: FilterProvider[]
  /** Context to be applied for all API queries */
  context?: Record<string, string>
  /** Variant of the filter button, whether to show the title or just the icon */
  filterButtonVariant?: FilterButtonVariant
  /** Initial value to use for the input. DO NOT USE in conjuction with `filterValue`.
   * Only use this if you want the Filter to have an initial value but remain uncontrolled. */
  initialFilterValue?: string
  /** Supplied filter value to use. By using this, the Filter is externally controlled
   * and will not use it's own state. */
  filterValue?: string
  /** React Ref to the input element */
  inputRef?: React.RefObject<HTMLInputElement>
  /** Label displayed atop the Filter bar. This is only shown when `visuallyHideLabel` is `false` */
  label: string
  /** Input keydown event callback */
  onKeyDown?: React.KeyboardEventHandler<Element>
  /** Validation callback, passing the validation message and the validation query */
  onValidation?: (messages: string[], filterQuery: FilterQuery) => void
  /** Placeholder text for the input */
  placeholder?: string
  /** Settings for the Filter, such as support of the legacy No filter provider and alias matching */
  settings?: FilterSettings
  /** Whether to show any validation messages or not. If this is disabled, use `onValidation` to receive the messages to display elsewhere */
  showValidationMessage?: boolean
  /** Variant of the Filter to render. `button` or `input` for one or the other, `full` (default) renders both.  */
  variant?: FilterVariant
  /** Whether to visually hide the Label or not */
  visuallyHideLabel?: boolean
} & TestIdProps &
  SxProp

const defaultFilterQuery = new FilterQuery()

export function Filter({
  context,
  'data-testid': dataTestId,
  filterButtonVariant = 'normal',
  filterValue,
  id,
  initialFilterValue,
  inputRef,
  label,
  onChange,
  onParse,
  onSubmit,
  onKeyDown,
  onValidation,
  placeholder,
  providers,
  settings,
  showValidationMessage = true,
  sx,
  variant = 'full',
  visuallyHideLabel = true,
}: FilterProps) {
  const [validationMessage, setValidationMessage] = useState<string[]>([])
  const hasValidationMessage = showValidationMessage && validationMessage.length > 0

  const validationCallback = useCallback(
    (messages: string[] = [], filterQuery: FilterQuery = defaultFilterQuery) => {
      setValidationMessage(messages)
      onValidation?.(messages, filterQuery)
    },
    [onValidation],
  )

  const filterButtonResponsive = useResponsiveValue<ResponsiveValue<FilterButtonVariant>, FilterButtonVariant>(
    {regular: filterButtonVariant, narrow: variant === 'button' ? 'normal' : 'compact'},
    filterButtonVariant,
  )

  const externallyControlled = typeof filterValue === 'string'
  const [innerValue, setInnerValue] = useState<string>(initialFilterValue || '')

  const changeCallback = useCallback(
    (value: string) => {
      if (!externallyControlled) {
        setInnerValue(value)
      }
      onChange?.(value)
    },
    [externallyControlled, onChange],
  )

  return (
    <FilterContextProvider
      context={context}
      inputRef={inputRef}
      rawFilter={externallyControlled ? filterValue : innerValue}
      onChange={changeCallback}
      onParse={onParse}
      onSubmit={onSubmit}
      onValidation={validationCallback}
      providers={providers}
      settings={{...defaultFilterSettings, ...settings}}
      variant={variant}
    >
      <Box
        id={id}
        role="form"
        className="FormControl FormControl--fullWidth"
        sx={{...sx, gap: 0}}
        {...testIdProps(dataTestId ?? 'filter')}
      >
        <label
          id={`${id}-label`}
          htmlFor={`${id}-input`}
          className={`FormControl-label ${visuallyHideLabel ? 'sr-only' : ''}`}
          {...testIdProps('filter-bar-label')}
        >
          {label}
        </label>
        <Box sx={{display: 'grid', gap: 3, width: '100%', position: 'relative'}}>
          <FilterInputWrapper isStandalone={variant === 'button'}>
            {variant !== 'input' && (
              <AdvancedFilterDialog isStandalone={variant === 'button'} filterButtonVariant={filterButtonResponsive} />
            )}
            {variant !== 'button' && (
              <>
                <Box
                  sx={filterInputWrapperStyles({
                    variant,
                    onSubmit: !!onSubmit,
                  })}
                >
                  {!onSubmit && <FilterInputIcon />}
                  <Input
                    id={id}
                    placeholder={placeholder}
                    onKeyDown={onKeyDown}
                    hasValidationMessage={hasValidationMessage}
                  />
                  <ClearButton />
                </Box>
                {onSubmit && <SubmitButton />}
              </>
            )}
          </FilterInputWrapper>
          <SuggestionsList id={id} />
          {/* This needs to move outside the Filter component because we can't predict where the consumer wants to put the error message. In memex this would go above the table view. */}
          {showValidationMessage && <ValidationMessage messages={validationMessage} id={`${id}-validation-message`} />}
        </Box>
      </Box>
    </FilterContextProvider>
  )
}
