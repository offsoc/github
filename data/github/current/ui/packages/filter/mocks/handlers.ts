import {delay, http, HttpResponse} from '@github-ui/storybook/msw'
import {hasMatch} from 'fzy.js'

import {caseInsensitiveStringCompare} from '../utils'
import {
  labels,
  labelSuggestions,
  languages,
  languageSuggestions,
  milestones,
  milestoneSuggestions,
  orgs,
  orgSuggestions,
  projects,
  projectSuggestions,
  repositories,
  repositorySuggestions,
  teams,
  teamSuggestions,
  users,
  userSuggestions,
} from './'

export const handlers = [
  [
    // #region Filter Suggestions
    // Labels
    http.get(`/_filter/labels`, async ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      await delay(500)
      if (!filterValue) {
        return HttpResponse.json({labels: labelSuggestions})
      }
      return HttpResponse.json({
        labels: labels.filter(l => {
          return hasMatch(filterValue, l.name)
        }),
      })
    }),
    // Languages
    http.get(`/_filter/languages`, ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) {
        return HttpResponse.json({languages: languageSuggestions})
      }
      return HttpResponse.json({
        languages: languages.filter(l => {
          return hasMatch(filterValue, l.name)
        }),
      })
    }),
    // Milestones
    http.get('/_filter/milestones', ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) {
        return HttpResponse.json({milestones: milestoneSuggestions})
      }
      return HttpResponse.json({
        milestones: milestoneSuggestions.filter(m => hasMatch(filterValue, m.title) || hasMatch(filterValue, m.value)),
      })
    }),
    // Orgs
    http.get('/_filter/organizations', ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) {
        return HttpResponse.json({orgs: orgSuggestions})
      }
      return HttpResponse.json({
        orgs: orgSuggestions.filter(o => hasMatch(filterValue, o.name)),
      })
    }),
    // Projects
    http.get(`/_filter/projects`, async ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) {
        return HttpResponse.json({projects: projectSuggestions})
      }

      await delay(500)
      return HttpResponse.json({
        projects: projects.filter(p => {
          return hasMatch(filterValue, p.title) || hasMatch(filterValue, p.value)
        }),
      })
    }),
    // Repositories
    http.get('/_filter/repositories', async ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) {
        return HttpResponse.json({repositories: repositorySuggestions})
      }

      await delay(500)
      return HttpResponse.json({
        repositories: repositorySuggestions.filter(r => hasMatch(filterValue, r.nameWithOwner)),
      })
    }),
    // Teams
    http.get(`/_filter/teams`, ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) {
        return HttpResponse.json({teams: teamSuggestions})
      }
      return HttpResponse.json({
        teams: teams.filter(t => {
          return hasMatch(filterValue, t.name) || hasMatch(filterValue, t.combinedSlug)
        }),
      })
    }),
    // Users
    http.get(`/_filter/users`, async ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) {
        return HttpResponse.json({users: userSuggestions})
      }

      await delay(500)
      return HttpResponse.json({
        users: users.filter(u => {
          return hasMatch(filterValue, u.login) || (u.name && hasMatch(filterValue, u.name))
        }),
      })
    }),
    //#endregion
    //#region Filter Validation
    // Labels
    http.get('/_filter/labels/validate', ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) return HttpResponse.json({message: 'Filter value is required'}, {status: 400})
      const label = labels.find(l => caseInsensitiveStringCompare(l.name, filterValue))
      if (!label) {
        return HttpResponse.json(
          {
            message: `Label '${filterValue}' does not exist`,
          },
          {status: 422},
        )
      }

      return HttpResponse.json(label)
    }),

    // Languages
    http.get(`/_filter/languages/validate`, ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) return HttpResponse.json({message: 'Filter value is required'}, {status: 400})
      const language = languages.find(l => caseInsensitiveStringCompare(l.name, filterValue))
      if (!language) {
        return HttpResponse.json(
          {
            message: `Language '${filterValue}' does not exist`,
          },
          {status: 422},
        )
      }

      return HttpResponse.json(language)
    }),
    // Milestones
    http.get('/_filter/milestones/validate', ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) return HttpResponse.json({message: 'Filter value is required'}, {status: 400})
      const milestone = milestones.find(m => caseInsensitiveStringCompare(m.title, filterValue))
      if (!milestone) {
        return HttpResponse.json(
          {
            message: `Milestone '${filterValue}' does not exist`,
          },
          {status: 422},
        )
      }

      return HttpResponse.json(milestone)
    }),
    // Orgs
    http.get('/_filter/organizations/validate', ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) return HttpResponse.json({message: 'Filter value is required'}, {status: 400})
      const org = orgs.find(o => caseInsensitiveStringCompare(o.name, filterValue))
      if (!org) {
        return HttpResponse.json(
          {
            message: `Org '${filterValue}' does not exist`,
          },
          {status: 422},
        )
      }

      return HttpResponse.json(org)
    }),
    // Projects
    http.get(`/_filter/projects/validate`, ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) return HttpResponse.json({message: 'Filter value is required'}, {status: 400})
      const project = projects.find(p => caseInsensitiveStringCompare(p.value, filterValue))
      if (!project) {
        return HttpResponse.json(
          {
            message: `Project '${filterValue}' does not exist`,
          },
          {status: 422},
        )
      }

      return HttpResponse.json(project)
    }),
    // Repositories
    http.get('/_filter/repositories/validate', ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) return HttpResponse.json({message: 'Filter value is required'}, {status: 400})
      const repository = repositories.find(r => caseInsensitiveStringCompare(r.nameWithOwner, filterValue))
      if (!repository) {
        return HttpResponse.json(
          {
            message: `Repository '${filterValue}' does not exist`,
          },
          {status: 422},
        )
      }

      return HttpResponse.json(repository)
    }),
    // Teams
    http.get(`/_filter/teams/validate`, ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')
      if (!filterValue) return HttpResponse.json({message: 'Filter value is required'}, {status: 400})
      const team = teams.find(t => caseInsensitiveStringCompare(t.combinedSlug, filterValue))
      if (!team) {
        return HttpResponse.json(
          {
            message: `Team '${filterValue}' does not exist`,
          },
          {status: 422},
        )
      }

      return HttpResponse.json(team)
    }),
    // Users
    http.get(`/_filter/users/validate`, ({request: req}) => {
      const filterValue = new URL(req.url, window.location.origin).searchParams.get('q')

      if (!filterValue) return HttpResponse.json({message: 'Filter value is required'}, {status: 400})
      // This is providing a false positive match
      // const user = filterValue === '@me' ? users[0] : users.find(u => u.login === filterValue)
      const user = users.find(u => caseInsensitiveStringCompare(u.login, filterValue))
      if (!user) {
        return HttpResponse.json(
          {
            message: `User '${filterValue}' does not exist`,
          },
          {status: 422},
        )
      }

      return HttpResponse.json(user)
    }),
  ],
]
