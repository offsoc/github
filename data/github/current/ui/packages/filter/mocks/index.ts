export * from './labels'
export * from './languages'
export * from './milestones'
export * from './orgs'
export * from './projects'
export * from './repositories'
export * from './teams'
export * from './users'
