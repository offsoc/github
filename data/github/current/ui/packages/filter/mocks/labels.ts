export const labelSuggestions = [
  {
    name: 'triage',
    nameHtml: `triage`,
    description: 'Triage',
    color: '#7a4dde',
  },
  {
    name: 'a11y',
    nameHtml: `a11y`,
    description: 'Accessibility',
    color: '#454ab0',
  },
  {
    name: 'accessibility review',
    nameHtml: `accessibility review`,
    description: 'Accessibility review',
    color: '#e668c3',
  },
  {
    name: 'batch',
    nameHtml: `batch`,
    description: 'Batch',
    color: '#69e011',
  },
  {
    name: 'availability',
    nameHtml: `availability`,
    description: 'Availability',
    color: '#d4111b',
  },
  {
    name: ':no_entry: blocked',
    nameHtml: `🚫 blocked`,
    description: 'No touchy!',
    color: '#79cb93',
  },
  {
    name: '🐛 bug',
    nameHtml: `🐛 bug`,
    description: 'Bug',
    color: '#44f3fa',
  },
  {
    name: '🚨 critical',
    nameHtml: `🚨 critical`,
    description: 'Critical',
    color: '#628e26',
  },
  {
    name: '🎨 design',
    nameHtml: `🎨 design`,
    description: 'Design',
    color: '#151f35',
  },
  {
    name: '🧪 experiment',
    nameHtml: `🧪 experiment`,
    description: 'Experiment',
    color: '#15146f',
  },
  {
    name: 'done',
    nameHtml: `done`,
    description: 'Done',
    color: '#bef627',
  },
  {
    name: '📝 documentation',
    nameHtml: `📝 documentation`,
    description: 'Documentation',
    color: '#f0c610',
  },
  {
    name: ':octocat: GitHub 😄',
    // This is test data and only used to mock an HTML return from the API
    // eslint-disable-next-line github/unescaped-html-literal
    nameHtml: `<img
        class="emoji"
        title=":octocat:"
        alt=":octocat:"
        src="https://github.githubassets.com/images/icons/emoji/octocat.png"
        height="20"
        width="20"
        align="absmiddle"
      />
      GitHub 😄`,
    description: 'The GitHub Octocats!',
    color: '#f00',
  },
  {
    name: '🔴 p1',
    nameHtml: `🔴 p1`,
    description: 'Priority 1',
    color: '#3fb4ca',
  },
  {
    name: '🟠 p2',
    nameHtml: `🟠 p2`,
    description: 'Priority 2',
    color: '#d8fb00',
  },
  {
    name: '🟡 p3',
    nameHtml: `🟡 p3`,
    description: 'Priority 3',
    color: '#bdbaf5',
  },
  {
    name: '🔵 p4',
    nameHtml: `🔵 p4`,
    description: 'Priority 4',
    color: '#074eb6',
  },
  {
    name: '🔒 security',
    nameHtml: `🔒 security`,
    description: 'Security',
    color: '#e313ca',
  },
  {
    name: 'tech debt',
    nameHtml: `tech debt`,
    description: 'Technical debt',
    color: '#75cf02',
  },
  {
    name: 'user story',
    nameHtml: `user story`,
    description: 'User story',
    color: '#dd6644',
  },
]

export const labels = [
  {
    name: 'triage',
    nameHtml: `triage`,
    description: 'Triage',
    color: '#7a4dde',
  },
  {
    name: 'a11y',
    nameHtml: `a11y`,
    description: 'Accessibility',
    color: '#454ab0',
  },
  {
    name: 'accessibility review',
    nameHtml: `accessibility review`,
    description: 'Accessibility review',
    color: '#e668c3',
  },
  {
    name: 'batch',
    nameHtml: `batch`,
    description: 'Batch',
    color: '#69e011',
  },
  {
    name: 'availability',
    nameHtml: `availability`,
    description: 'Availability',
    color: '#d4111b',
  },
  {
    name: ':no_entry: blocked',
    nameHtml: `🚫 blocked`,
    description: 'No touchy!',
    color: '#79cb93',
  },
  {
    name: '🐛 bug',
    nameHtml: `🐛 bug`,
    description: 'Bug',
    color: '#44f3fa',
  },
  {
    name: '🚨 critical',
    nameHtml: `🚨 critical`,
    description: 'Critical',
    color: '#628e26',
  },
  {
    name: '🎨 design',
    nameHtml: `🎨 design`,
    description: 'Design',
    color: '#151f35',
  },
  {
    name: '🧪 experiment',
    nameHtml: `🧪 experiment`,
    description: 'Experiment',
    color: '#15146f',
  },
  {
    name: 'done',
    nameHtml: `done`,
    description: 'Done',
    color: '#bef627',
  },
  {
    name: '📝 documentation',
    nameHtml: `📝 documentation`,
    description: 'Documentation',
    color: '#f0c610',
  },
  {
    name: ':octocat: GitHub 😄',
    // This is test data and only used to mock an HTML return from the API
    // eslint-disable-next-line github/unescaped-html-literal
    nameHtml: `<img
        class="emoji"
        title=":octocat:"
        alt=":octocat:"
        src="https://github.githubassets.com/images/icons/emoji/octocat.png"
        height="20"
        width="20"
        align="absmiddle"
      />
      GitHub 😄`,
    description: 'The GitHub Octocats!',
    color: '#f00',
  },
  {
    name: '🔴 p1',
    nameHtml: `🔴 p1`,
    description: 'Priority 1',
    color: '#3fb4ca',
  },
  {
    name: '🟠 p2',
    nameHtml: `🟠 p2`,
    description: 'Priority 2',
    color: '#d8fb00',
  },
  {
    name: '🟡 p3',
    nameHtml: `🟡 p3`,
    description: 'Priority 3',
    color: '#bdbaf5',
  },
  {
    name: '🔵 p4',
    nameHtml: `🔵 p4`,
    description: 'Priority 4',
    color: '#074eb6',
  },
  {
    name: '🔒 security',
    nameHtml: `🔒 security`,
    description: 'Security',
    color: '#e313ca',
  },
  {
    name: 'tech debt',
    nameHtml: `tech debt`,
    description: 'Technical debt',
    color: '#75cf02',
  },
  {
    name: 'user story',
    nameHtml: `user story`,
    description: 'User story',
    color: '#dd6644',
  },
  {
    name: 'team:engineering',
    nameHtml: `team:engineering`,
    description: 'Team: Engineering',
    color: '#074eb6',
  },
  {
    name: 'a11y (sev 1)',
    nameHtml: `a11y (sev 1)`,
    description: 'Accessibility (Severity 1)',
    color: '#000000',
  },
  {
    name: '(cool',
    nameHtml: `(cool`,
    description: '(cool',
    color: '#a2eeef',
  },
]
