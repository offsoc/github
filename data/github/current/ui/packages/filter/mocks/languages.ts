export const languageSuggestions = [
  {
    name: 'C++',
    color: '#f34b7d',
  },
  {
    name: 'C#',
    color: '#178600',
  },
  {
    name: 'Go',
    color: '#00ADD8',
  },
  {
    name: 'Java',
    color: '#b07219',
  },
  {
    name: 'JavaScript',
    color: '#f1e05a',
  },
  {
    name: 'PHP',
    color: '#4F5D95',
  },
  {
    name: 'Python',
    color: '#3572A5',
  },
  {
    name: 'Ruby',
    color: '#701516',
  },
]

export const languages = [
  {
    name: 'C++',
    color: '#f34b7d',
  },
  {
    name: 'C#',
    color: '#178600',
  },
  {
    name: 'Go',
    color: '#00ADD8',
  },
  {
    name: 'Java',
    color: '#b07219',
  },
  {
    name: 'JavaScript',
    color: '#f1e05a',
  },
  {
    name: 'PHP',
    color: '#4F5D95',
  },
  {
    name: 'Python',
    color: '#3572A5',
  },
  {
    name: 'Ruby',
    color: '#701516',
  },
]
