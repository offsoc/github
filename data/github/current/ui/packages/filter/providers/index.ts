export {AsyncFilterProvider} from './async'
export * from './key-only'
export {LabelFilterProvider} from './label'
export {LanguageFilterProvider} from './language'
export {MilestoneFilterProvider} from './milestone'
export {OrgFilterProvider} from './org'
export {ProjectFilterProvider} from './project'
export {RepositoryFilterProvider} from './repository'
export * from './static'
export {TeamFilterProvider, TeamReviewRequestedFilterProvider} from './team'
export {TypeFilterProvider} from './type'
export * from './user'
