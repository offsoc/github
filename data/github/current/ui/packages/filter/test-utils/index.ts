import {act, fireEvent, screen} from '@testing-library/react'

export const setupExpectedAsyncErrorHandler = () => {
  jest.spyOn(console, 'error').mockImplementation((message: string) => {
    // * Because Filter is asynchronous, there are console errors that are thrown, but expected. This will rethrow
    // * any errors that are not related to the async nature of the component.
    if (!message.includes?.('wrapped in act(')) {
      // eslint-disable-next-line no-console
      console.warn(message)
    }
  })
}

export async function updateFilterValue(key: string, value: string = '') {
  jest.useFakeTimers()

  const input = screen.getByRole('combobox')
  input.focus()

  fireEvent.change(input, {target: {value: `${key}${key && value ? ':' : ''}${value}`}})

  // eslint-disable-next-line @typescript-eslint/require-await
  await act(async () => {
    jest.runAllTimers()
  })
}
