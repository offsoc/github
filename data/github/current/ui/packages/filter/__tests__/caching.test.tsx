import {isFeatureEnabled} from '@github-ui/feature-flags'
import {render} from '@github-ui/react-core/test-utils'
import {act} from '@testing-library/react'
import {hasMatch} from 'fzy.js'

import {Filter} from '../Filter'
import {labels, labelSuggestions} from '../mocks'
import {LabelFilterProvider} from '../providers'
import {setupExpectedAsyncErrorHandler, updateFilterValue} from '../test-utils'
import {
  appendToFilterAndRenderAsyncSuggestions,
  expectFilterValueToBe,
  expectSuggestionsToBeEmpty,
  expectSuggestionsToMatchSnapshot,
  selectSuggestion,
} from './utils/helpers'

jest.mock('@github-ui/feature-flags')

jest.setTimeout(4_500)
const globalFetch = global.fetch

describe('Caching suggestions', () => {
  const set = new Set()
  let error: Error | null = null

  beforeEach(() => {
    global.fetch = jest.fn(url => {
      const parsedUrl = new URL(url, window.location.origin)
      const filterValue = parsedUrl.searchParams.get('q')
      // fail the test if fetch is called more than once for the same value
      if (set.has(url)) {
        error = new Error('Error: fetch should not be called more than once for value')
        return Promise.reject(error)
      }
      set.add(url)
      return Promise.resolve({
        ok: true,
        json: () =>
          Promise.resolve({
            labels: filterValue
              ? labels.filter(l => {
                  return hasMatch(filterValue, l.name)
                })
              : labelSuggestions,
          }),
      })
    }) as jest.Mock
    setupExpectedAsyncErrorHandler()
  })

  afterEach(() => {
    set.clear()
    jest.restoreAllMocks()
    ;(isFeatureEnabled as jest.Mock).mockRestore()
    global.fetch = globalFetch
  })

  function expectFetchToBeCalledOnce() {
    if (error) throw error
  }

  it('should filter and select cached suggestions on empty value', async () => {
    const filterProviders = [new LabelFilterProvider()]
    const {user} = render(<Filter id="test-filter-bar" label="Filter" providers={filterProviders} />)

    await updateFilterValue('label:')
    expectFetchToBeCalledOnce()

    expectSuggestionsToMatchSnapshot()

    await appendToFilterAndRenderAsyncSuggestions('a')
    expectFetchToBeCalledOnce()

    expectSuggestionsToMatchSnapshot()

    await user.keyboard('{Backspace}')

    // eslint-disable-next-line @typescript-eslint/require-await
    await act(async () => {
      jest.runAllTimers()
    })
    expectFetchToBeCalledOnce()

    expectSuggestionsToMatchSnapshot()

    await selectSuggestion('No label')
    expectFilterValueToBe('no:label')
  })

  it('should filter and select cached suggestions on value', async () => {
    const filterProviders = [new LabelFilterProvider()]
    render(<Filter id="test-filter-bar" label="Filter" providers={filterProviders} />)

    await updateFilterValue('label:a')
    expectFetchToBeCalledOnce()

    expectSuggestionsToMatchSnapshot()

    await selectSuggestion('triage')

    expectFilterValueToBe('label:triage')

    await appendToFilterAndRenderAsyncSuggestions(',b')
    expectFetchToBeCalledOnce()

    expectSuggestionsToMatchSnapshot()

    await selectSuggestion('accessibility review')

    expectFilterValueToBe('label:triage,"accessibility review"')
  })

  it('should filter and select cached suggestions on value with prefetching enabled', async () => {
    ;(isFeatureEnabled as jest.Mock).mockImplementation(flag => flag === 'filter_prefetch_suggestions')

    const filterProviders = [new LabelFilterProvider()]
    render(<Filter id="test-filter-bar" label="Filter" providers={filterProviders} />)

    await updateFilterValue('label:a')
    expectFetchToBeCalledOnce()

    expectSuggestionsToMatchSnapshot()

    await selectSuggestion('triage')

    expectFilterValueToBe('label:triage')

    await appendToFilterAndRenderAsyncSuggestions(',b')
    expectFetchToBeCalledOnce()

    expectSuggestionsToMatchSnapshot()

    await selectSuggestion('accessibility review')

    expectFilterValueToBe('label:triage,"accessibility review"')
  })

  it('should filter and select cached suggestions on backspace', async () => {
    const filterProviders = [new LabelFilterProvider()]
    const {user} = render(<Filter id="test-filter-bar" label="Filter" providers={filterProviders} />)

    await updateFilterValue('label:')
    expectFetchToBeCalledOnce()

    await appendToFilterAndRenderAsyncSuggestions('a11')
    expectFetchToBeCalledOnce()

    await appendToFilterAndRenderAsyncSuggestions('y')
    expectFetchToBeCalledOnce()

    await expectSuggestionsToBeEmpty()

    await user.keyboard('{Backspace>4}')
    expectFetchToBeCalledOnce()

    expectFilterValueToBe('label:')
  })

  it('should not display selected from cached suggestion', async () => {
    const filterProviders = [new LabelFilterProvider()]
    render(<Filter id="test-filter-bar" label="Filter" providers={filterProviders} />)

    await updateFilterValue('label:c')

    expectFetchToBeCalledOnce()

    expectSuggestionsToMatchSnapshot()

    await selectSuggestion('accessibility review')

    expectFilterValueToBe('label:"accessibility review"')

    await appendToFilterAndRenderAsyncSuggestions(',c')

    expectFetchToBeCalledOnce()

    expectSuggestionsToMatchSnapshot()

    await selectSuggestion('batch')

    expectFilterValueToBe('label:"accessibility review",batch')
  })
})
