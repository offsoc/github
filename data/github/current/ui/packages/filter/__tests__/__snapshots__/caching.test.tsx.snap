// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Caching suggestions should filter and select cached suggestions on empty value 1`] = `
[
  "No label",
  "Exclude label",
  "triage",
  "a11y",
  "accessibility review",
  "batch",
  "availability",
  "🚫 blocked",
  "🐛 bug",
  "🚨 critical",
  "🎨 design",
  "🧪 experiment",
  "done",
  "📝 documentation",
  "
      GitHub 😄",
  "🔴 p1",
  "🟠 p2",
  "🟡 p3",
  "🔵 p4",
  "🔒 security",
  "tech debt",
  "user story",
]
`;

exports[`Caching suggestions should filter and select cached suggestions on empty value 2`] = `
[
  "a11y",
  "availability",
  "a11y (sev 1)",
  "accessibility review",
  "batch",
  "triage",
  "🚨 critical",
  "📝 documentation",
  "team:engineering",
  "
      GitHub 😄",
]
`;

exports[`Caching suggestions should filter and select cached suggestions on empty value 3`] = `
[
  "No label",
  "Exclude label",
  "triage",
  "a11y",
  "accessibility review",
  "batch",
  "availability",
  "🚫 blocked",
  "🐛 bug",
  "🚨 critical",
  "🎨 design",
  "🧪 experiment",
  "done",
  "📝 documentation",
  "
      GitHub 😄",
  "🔴 p1",
  "🟠 p2",
  "🟡 p3",
  "🔵 p4",
  "🔒 security",
  "tech debt",
  "user story",
]
`;

exports[`Caching suggestions should filter and select cached suggestions on value 1`] = `
[
  "a11y",
  "availability",
  "a11y (sev 1)",
  "accessibility review",
  "batch",
  "triage",
  "🚨 critical",
  "📝 documentation",
  "team:engineering",
  "
      GitHub 😄",
]
`;

exports[`Caching suggestions should filter and select cached suggestions on value 2`] = `
[
  "batch",
  "🐛 bug",
  "🚫 blocked",
  "tech debt",
  "availability",
  "
      GitHub 😄",
  "accessibility review",
]
`;

exports[`Caching suggestions should filter and select cached suggestions on value with prefetching enabled 1`] = `
[
  "a11y",
  "availability",
  "a11y (sev 1)",
  "accessibility review",
  "batch",
  "triage",
  "🚨 critical",
  "📝 documentation",
  "team:engineering",
  "
      GitHub 😄",
]
`;

exports[`Caching suggestions should filter and select cached suggestions on value with prefetching enabled 2`] = `
[
  "batch",
  "🐛 bug",
  "🚫 blocked",
  "tech debt",
  "availability",
  "
      GitHub 😄",
  "accessibility review",
]
`;

exports[`Caching suggestions should not display selected from cached suggestion 1`] = `
[
  "🚨 critical",
  "batch",
  "(cool",
  "tech debt",
  "🔒 security",
  "📝 documentation",
  "🚫 blocked",
  "
      GitHub 😄",
  "accessibility review",
]
`;

exports[`Caching suggestions should not display selected from cached suggestion 2`] = `
[
  "🚨 critical",
  "batch",
  "(cool",
  "tech debt",
  "🔒 security",
  "📝 documentation",
  "🚫 blocked",
  "
      GitHub 😄",
]
`;
